File: bindings/python/examples/memory_scanner.py
"""
Example: Memory scanner with Python
"""

import inspectra

def main():
    inspectra.init()
    
    print("Inspectra Memory Scanner Example\n")
    
    # Get target process
    pid = int(input("Enter process PID: "))
    
    try:
        scanner = inspectra.Scanner(pid)
        print(f"Attached to process: {pid}\n")
        
        while True:
            value_input = input("Enter value to scan (or 'quit'): ")
            
            if value_input.lower() == 'quit':
                break
            
            try:
                value = int(value_input)
                results = scanner.scan_i32(value)
                
                print(f"Found {len(results)} results")
                
                # Display first 10
                for i, addr in enumerate(results[:10]):
                    print(f"  [{i}] Address: 0x{addr:X}")
                
                if len(results) > 10:
                    print(f"  ... and {len(results) - 10} more")
                
            except ValueError:
                print("Invalid value")
    
    except Exception as e:
        print(f"Error: {e}")
    
    print("Exiting...")

if __name__ == "__main__":
    main()

File: core/examples/memory_scanner.rs
//! Example: Memory scanner

use inspectra_core::{memory, process, scanner, types};
use scanner::{ScanConfig, Scanner};
use std::io::{self, Write};
use types::DataType;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    inspectra_core::init()?;

    println!("Inspectra Memory Scanner Example\n");

    // Get target process
    print!("Enter process name or PID: ");
    io::stdout().flush()?;

    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    let input = input.trim();

    let manager = process::get_process_manager();

    // Try to parse as PID, otherwise search by name
    let handle = if let Ok(pid) = input.parse::<u32>() {
        manager.attach(pid)?
    } else {
        let procs = manager.find_by_name(input)?;
        if procs.is_empty() {
            eprintln!("No process found with name: {}", input);
            return Ok(());
        }
        println!(
            "Found {} processes, using first: {} (PID: {})",
            procs.len(),
            procs[0].name,
            procs[0].pid
        );
        manager.attach(procs[0].pid)?
    };

    println!("Attached to process: PID {}", handle.pid());

    // Create memory accessor
    let mem = memory::create_memory(handle.as_ref())?;

    // Configure scanner
    let config = ScanConfig {
        data_type: DataType::I32,
        ..Default::default()
    };

    let mut scanner = Scanner::new(mem, config);

    // Scan loop
    loop {
        print!("\nEnter value to scan for (or 'quit'): ");
        io::stdout().flush()?;

        let mut value_input = String::new();
        io::stdin().read_line(&mut value_input)?;
        let value_input = value_input.trim();

        if value_input.eq_ignore_ascii_case("quit") {
            break;
        }

        if let Ok(value) = value_input.parse::<i32>() {
            println!("Scanning for value: {}", value);

            let bytes = value.to_le_bytes();
            let results = scanner.scan(Some(&bytes), None)?;

            println!("Found {} results", results.len());

            // Display first 10 results
            for (i, result) in results.iter().take(10).enumerate() {
                println!("  [{}] Address: 0x{:X}", i, result.address);
            }

            if results.len() > 10 {
                println!("  ... and {} more", results.len() - 10);
            }
        } else {
            println!("Invalid value");
        }
    }

    println!("Exiting...");
    Ok(())
}
